### PR TITLE
feat(python): Improve consistency of `dtype` inference from Python types

### LINKED ISSUE #13117: Consistently support Python types as aliases for polars types
### Description

Polars sometimes supports Python types (e.g. `str`, `float`, `int`) as aliases for polars types, and sometimes not.

For instance, these are inconsistent:

```python
>>> pl.Series([1], dtype=pl.Float64)
shape: (1,)
Series: '' [f64]
[
        1.0
]
>>> pl.Series([1], dtype=float)
shape: (1,)
Series: '' [i64]
[
        1
]
```

as are these:

```python
>>> pl.DataFrame({'a': [1]}).select(pl.all().cast(str))
 a
 1
shape: (1, 1)
>>> pl.DataFrame({'a': [1]}).cast(str)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/wainberg/miniforge3/lib/python3.12/site-packages/polars/dataframe/frame.py", line 6756, in cast
    return self.lazy().cast(dtypes, strict=strict).collect(_eager=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wainberg/miniforge3/lib/python3.12/site-packages/polars/lazyframe/frame.py", line 2446, in cast
    return self._from_pyldf(self._ldf.cast_all(dtypes, strict))
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: argument 'dtype': 'type' is not a Polars data type
```

and these:

```python
>>> pl.DataFrame({'a': ['foo']}).select(pl.col(pl.Utf8))
 a
 foo
shape: (1, 1)
>>> pl.DataFrame({'a': ['foo']}).select(pl.col(str))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/wainberg/miniforge3/lib/python3.12/site-packages/polars/functions/col.py", line 287, in __new__
    return _create_col(name, *more_names)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wainberg/miniforge3/lib/python3.12/site-packages/polars/functions/col.py", line 61, in _create_col
    raise TypeError(
TypeError: invalid input for `col`

Expected `str` or `DataType`, got 'type'.
```

### LINKED ISSUE #14599: Wrong polars dtype for empty Series
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python

```
>>> import polars as pl
>>> ser1 = pl.Series('str', dtype=str)
>>> ser2 = pl.Series('int', dtype=pl.Int32)
>>> print(ser1, ser2)
shape: (0,)
Series: 'str' [null]
[
] shape: (0,)
Series: 'int' [i32]
[
]

### Log output

_No response_

### Issue description

According to the API reference, ser1 should have dtype [str] since I have appointed it. Nevertheless, the dtype remains [null]. However, ser2 has the right dtype.

### Expected behavior

>>> import polars as pl
>>> ser = pl.Series('str', dtype=str)
>>> print(ser)
shape: (0,)
Series: 'str' [str]
[
]

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.10
Index type:           UInt32
Platform:             Linux-3.10.0-1160.90.1.el7.x86_64-x86_64-with-glibc2.18
Python:               3.12.2 | packaged by conda-forge | (main, Feb 16 2024, 20:50:58) [GCC 12.3.0]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fsspec:               2023.10.0
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           <not installed>
numpy:                1.26.4
openpyxl:             <not installed>
pandas:               2.2.0
pyarrow:              15.0.0
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           2.0.27
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by alexander-beedie**
Largely addresses #13117 (fixing dtype treatment on the Python side), closes #14599.

Consolidates some dtype inference (where we would accept python types), and improves numpy dtype detection (specifically: datetime, duration, binary dtypes).

### CODE REVIEW

**Comment by alexander-beedie** [comment]
> I don't think it's a good idea to support NumPy type inputs for our `dtype` parameters. Supporting `str/int/float`, sure. But I don't want to support dtypes from other libraries. Should we then also support pandas / pyarrow inputs? I'd rather not go there.

True; I was 50/50 on that - will remove the two lines that trigger it, while leaving the improved numpy dtype detection in 👌

File: py-polars/polars/series/series.py [hunk 1]
```diff
@@ -268,17 +268,18 @@ def __init__(
...
-            dtype is not None
-            and not is_polars_dtype(dtype)
-            and py_type_to_dtype(dtype, raise_unmatched=False) is None
-        ):
-            msg = f"given dtype: {dtype!r} is not a valid Polars data type and cannot be converted into one"
-            raise ValueError(msg)
+        elif dtype is not None and not is_polars_dtype(dtype):
+            # Raise early error on invalid dtype
+            if not is_polars_dtype(
+                py_dtype := py_type_to_dtype(dtype, raise_unmatched=False)
```
Comment (stinodego) [inline py-polars/polars/series/series.py:277]: nitpicky: the variable `py_dtype` is a polars data type, so can we name this `pl_dtype` instead?

File: py-polars/polars/datatypes/convert.py [hunk 2]
```diff
@@ -464,11 +466,23 @@ def supported_numpy_char_code(dtype_char: str) -> bool:
...
 
 
+def numpy_type_to_dtype(dtype: np.dtype) -> PolarsDataType:  # type: ignore[type-arg]
+    """Convert a numpy dtype to a Polars dtype."""
+    char_code = np.dtype(dtype).char
+    pl_dtype = numpy_char_code_to_dtype(char_code)
+    if pl_dtype in (Datetime, Duration):
+        if (unit := dtype.str[-3:-1]) in ("ms", "us", "ns"):
+            pl_dtype = Datetime(unit) if dtype.kind == "M" else Duration(unit)  # type: ignore[arg-type]
+    return pl_dtype
```
Comment (stinodego) [inline py-polars/polars/datatypes/convert.py:476]: Looks like this isn't used anywhere? Better not to include it then.

File: py-polars/polars/series/series.py [hunk 3]
```diff
@@ -268,17 +268,18 @@ def __init__(
...
+            ):
+                msg = f"given dtype: {dtype!r} is not a valid Polars data type and cannot be converted into one"
+                raise ValueError(msg)
+            else:
+                dtype = py_dtype
```
Comment (stinodego) [inline py-polars/polars/series/series.py:282]: Good idea to just use the inferred dtype here 👍 we've already done the work.

**Review by stinodego — CHANGES_REQUESTED** [review]
All right, some small comments then this can go in.

File: py-polars/polars/series/series.py (hunk 1 above)
Comment (alexander-beedie) [reply py-polars/polars/series/series.py:277]: Good spot; in my head I was thinking `from_py_dtype` and omitted the leading `from_` 😅

**Review by stinodego — APPROVED** [review]
Thanks!

**Pull request merged**